geo/src/algorithm/bool_ops/assembly.rs

```diff
@@ -0,0 +1,335 @@
+                }
+            }
+        }
+
+        let (rings, snakes_idx_map) = rings_from_snakes(&mut snakes[..]);
+
+        let mut polygons = vec![];
+        let mut children = HashMap::new();
+        for (ring_idx, ring) in rings.iter().enumerate() {
+            if ring.is_hole {
+                let mut parent_ring_idx;
+                let mut parent_snake_idx = ring.parent_snake_idx;
+                loop {
+                    parent_ring_idx = snakes_idx_map[&parent_snake_idx];
+                    let parent = &rings[parent_ring_idx];
+                    if !parent.is_hole {
+                        break;
+                    }
+                    parent_snake_idx = parent.parent_snake_idx;
+                }
+                let this_children = children.entry(parent_ring_idx).or_insert(vec![]);
+                this_children.push(ring_idx);
+            } else {
+                continue;
+            }
+        }
+
+        for (ring_idx, ring) in rings.iter().enumerate() {
+            if ring.is_hole {
+                continue;
+            }
+            let mut holes = vec![];
+            for child_idx in children.remove(&ring_idx).unwrap_or_default() {
+                let ls = split_ring(&rings[child_idx].ls, |ls| holes.push(ls));
+                holes.push(ls);
+            }
+            debug!("ext: {ext:?}", ext = ring.ls);
+            let exterior = split_ring(&ring.ls, |ls| holes.push(ls));
+            polygons.push(Polygon::new(exterior, holes));
+        }
+
+        polygons.into()
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Ring<T: GeoFloat> {
+    ls: LineString<T>,
+    is_hole: bool,
+    parent_snake_idx: usize,
+}
+
+fn split_ring<T: GeoFloat, F: FnMut(LineString<T>)>(
+    ls: &LineString<T>,
+    mut cb: F,
+) -> LineString<T> {
+    let mut pts_map = BTreeMap::new();
+    let mut exterior = vec![];
+    for coord in ls.0.iter().copied() {
+        if let Some(idx) = pts_map.get(&SweepPoint::from(coord)) {
+            let new_ls: LineString<_> = exterior
+                .drain(idx..)
+                .inspect(|pt| {
+                    pts_map.remove(&SweepPoint::from(*pt)).unwrap();
+                })
+                .collect();
+            cb(new_ls);
+        }
+        pts_map.insert(SweepPoint::from(coord), exterior.len());
+        exterior.push(coord);
+    }
+    pts_map.clear();
+    LineString::from(exterior)
+}
+
+fn rings_from_snakes<T: GeoFloat>(
+    snakes: &mut [Snake<T>],
+) -> (Vec<Ring<T>>, HashMap<usize, usize>) {
+    let mut snake_idx_map = HashMap::new();
+    let mut rings = vec![];
+    for idx in 0..snakes.len() {
+        if let Some(ls) = Snake::into_ring(snakes, idx, |midx| {
+            snake_idx_map.insert(midx, rings.len());
+        }) {
+            rings.push(ls);
+        }
+    }
+    (rings, snake_idx_map)
+}
+
+#[derive(Debug, Clone)]
+struct Snake<T: GeoFloat> {
+    points: Vec<SweepPoint<T>>,
+    start_pair: usize,
+    end_pair: Option<usize>,
```

Comment:
`start_pair` is the snake_idx of the snake that preceded this snake in the ring construction.
`end_pair` is the snake_idx of the snake that comes *after* this snake in the ring construction.

Is that right?

Reply:
The topology guarantees that snakes start out and end in pairs (but they may not be the same pair).  The indices are needed to finally join the snakes into a ring.  At the point of ring construction, we'll have a bunch of points, and snakes connecting different points. This is a graph, and is Eulerian (number of snakes start/ending at each point is even), so we connect them into a set of rings.

There are some additional constraints on the output topology:  for instance two polygons touching at a vertex shouldn't be merged into one; or hole touching the exterior of a polygon shouldn't me merged into the exterior.  This requires additional info. stored in the snake.